src/MainWindow.cpp

```diff
@@ -196,7 +196,17 @@ bool MainWindowPrivate::validateCardError( QSmartCardData::PinType type, int fla
 	switch( err )
 	{
 	case QSmartCard::NoError: return true;
-	case QSmartCard::CancelError: break;
+	case QSmartCard::CancelError:
+		if ( q->d->smartcard->data().isPinpad() )
+		{
+			switch ( type )
+			{
+			case QSmartCardData::Pin1Type: showWarning( tr("PIN1 timeout") ); break;
+			case QSmartCardData::Pin2Type: showWarning( tr("PIN2 timeout") ); break;
+			case QSmartCardData::PukType: showWarning( tr("PUK timeout") ); break;
+			}
+		}
+		break;
```

Comment:
It was a main requirement of IB-4807. Same way it works in qDigidoc3 client. I suppose that both applications should work the same way. Also user can cancel operation (progress bar), but it continue to work in background until time limit and then the timeout message is shown, which is OK to my mind. Tells user that she/he can continue using application.